dGame/dCinema/Recorder.cpp

```diff
@@ -0,0 +1,1195 @@
+		records.push_back(record);
+
+		if (element->Attribute("name")) {
+			records.back()->m_Name = element->Attribute("name");
+		}
+	}
+
+}
+
+Entity* Recorder::ActFor(Entity* actorTemplate, Entity* player, Play* variables) {
+	EntityInfo info;
+	info.lot = actorTemplate->GetLOT();
+	info.pos = actorTemplate->GetPosition();
+	info.rot = actorTemplate->GetRotation();
+	info.scale = 1;
+	info.spawner = nullptr;
+	info.spawnerID = player->GetObjectID();
+	info.spawnerNodeID = 0;
+	info.settings = {
+		new LDFData<std::vector<std::u16string>>(u"syncLDF", { u"custom_script_client" }),
+		new LDFData<std::u16string>(u"custom_script_client", u"scripts\\ai\\SPEC\\MISSION_MINIGAME_CLIENT.lua")
+	};
+
+	// Spawn it
+	auto* actor = Game::entityManager->CreateEntity(info);
+
+	// Hide the template from the player
+	ServerPreconditions::AddExcludeFor(player->GetObjectID(), actorTemplate->GetObjectID());
+
+	// Solo act for the player
+	ServerPreconditions::AddSoloActor(actor->GetObjectID(), player->GetObjectID());
+
+	Game::entityManager->ConstructEntity(actor);
+
+	Act(actor, variables);
+
+	return actor;
+}
+
+void Recorder::StopActingFor(Entity* actor, Entity* actorTemplate, LWOOBJID playerID) {
+	// Remove the exclude for the player
+	ServerPreconditions::RemoveExcludeFor(playerID, actorTemplate->GetObjectID());
+
+	Game::entityManager->DestroyEntity(actor);
+}
+
+bool Recorder::IsRecording() const {
+	return this->m_IsRecording;
+}
+
+void Recorder::StartRecording(LWOOBJID actorID) {
+	const auto& it = m_Recorders.find(actorID);
+
+	// Delete the old recorder if it exists
+	if (it != m_Recorders.end()) {
+		delete it->second;
+		m_Recorders.erase(it);
+	}
+
+	Recorder* recorder = new Recorder();
+	m_Recorders.insert_or_assign(actorID, recorder);
+	recorder->m_IsRecording = true;
+}
+
+void Recorder::StopRecording(LWOOBJID actorID) {
+	auto iter = m_Recorders.find(actorID);
+	if (iter == m_Recorders.end()) {
+		return;
+	}
+
+	iter->second->m_IsRecording = false;
+}
+
+Recorder* Recorder::GetRecorder(LWOOBJID actorID) {
+	auto iter = m_Recorders.find(actorID);
+	if (iter == m_Recorders.end()) {
+		return nullptr;
+	}
+
+	return iter->second;
+}
+
+void Cinema::Recording::Recorder::RegisterEffectForActor(LWOOBJID actorID, const int32_t& effectId) {
+	auto iter = m_Recorders.find(actorID);
+	if (iter == m_Recorders.end()) {
+		return;
+	}
+
+	auto& recorder = iter->second;
+
+	const auto& effectIter = m_EffectAnimations.find(effectId);
+
+	if (effectIter == m_EffectAnimations.end()) {
+		auto statement = CDClientDatabase::CreatePreppedStmt("SELECT animationName FROM BehaviorEffect WHERE effectID = ? LIMIT 1;");
+
+		statement.bind(1, effectId);
+
+		auto result = statement.execQuery();
+
+		if (result.eof()) {
+			result.finalize();
+
+			m_EffectAnimations.emplace(effectId, "");
+		}
+		else {
+			const auto animationName = result.getStringField(0);
+
+			m_EffectAnimations.emplace(effectId, animationName);
+
+			result.finalize();
```

Comment:
finalize isnt necessary to manually call, the destructor for resultset here already calls finalize in a try {finalize()} catch (...) {};